src/compilation/options.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{
+    compilation::{compile_validators, context::CompilationContext, JSONSchema, DEFAULT_SCOPE},
+    error::CompilationError,
+    resolver::Resolver,
+    schemas,
+};
+use serde_json::Value;
+use std::borrow::Cow;
+
+#[derive(Clone, Debug, Default)]
+pub struct CompilationOptions {
+    pub(crate) draft: Option<schemas::Draft>,
+}
+
+impl CompilationOptions {
+    pub(crate) fn draft(&self) -> schemas::Draft {
+        self.draft.unwrap_or_default()
+    }
+
+    /// Compile `schema` into `JSONSchema` using the currently defined options.Draft
+    pub fn compile<'a>(&self, schema: &'a Value) -> Result<JSONSchema<'a>, CompilationError> {
+        // Draft is detected in the following precedence order:
+        //   - Explicitly specified;
+        //   - $schema field in the document;
+        //   - Draft::default()
+
+        // Clone needed because we are going to store a Copy-on-Write (Cow) instance
+        // into the final JSONSchema as well as passing `self` (the instance and not
+        // the reference) would require Copy trait implementation from
+        // `CompilationOptions` which is something that we would like to avoid as
+        // options might contain heap-related objects (ie. an HashMap) and we want the
+        // memory-related operations to be explicit
+        let mut config = self.clone();
+        if self.draft.is_none() {
+            if let Some(draft) = schemas::draft_from_schema(schema) {
+                config.draft = Some(draft);
+            }
+        }
+        let processed_config: Cow<'_, CompilationOptions> = Cow::Owned(config);
+        let draft = processed_config.draft();
+
+        let scope = match schemas::id_of(draft, schema) {
+            Some(url) => url::Url::parse(url)?,
+            None => DEFAULT_SCOPE.clone(),
+        };
+        let resolver = Resolver::new(draft, &scope, schema)?;
+        let context = CompilationContext::new(scope, processed_config);
+
+        let mut validators = compile_validators(schema, &context)?;
+        validators.shrink_to_fit();
+
+        Ok(JSONSchema {
+            schema,
+            resolver,
+            validators,
+            context,
+        })
+    }
+
+    pub fn set_draft(&mut self, draft: schemas::Draft) -> &mut Self {
```

Comment:
As you noted we do need to have the getter and so only one of them can be named `draft`.
I was trying to follow the [conventions for setter methods (rfc-344)](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344]).

But I'm open to suggestions
| getter | setter | description |
|--|--|--|
| draft | set_draft | current in PR |
| draft | with_draft |  |
| draft | froce_draft | |
| draft_or_default | draft | it would be very long and expose the user of the API to the internal behaviour (sort of contract) |

Reply:
Indeed, good note!

I like the draft/with_draft variant